---

## Part 0｜Setup - 讀取資料

### 🎯 任務說明

首先，我們要把資料讀進來，並做初步觀察。

**請完成以下步驟：**
1. import pandas
2. 讀取 `data/dirty_orders.csv`
3. 觀察資料的前幾筆、後幾筆、隨機幾筆
4. 確認資料的大小（幾列、幾欄）
5. 列出所有欄位名稱

---

### 📌 提示
- 使用 `pd.read_csv()`
- 使用 `df.head()`, `df.tail()`, `df.sample()`
- 使用 `df.shape` 和 `df.columns`

In [1]:
# TODO: import pandas as pd
import pandas as pd

In [25]:
# TODO: 讀取 data/dirty_orders.csv
# df = pd.read_csv(...)

df = pd.read_csv('./data/dirty_orders.csv', on_bad_lines='skip')

df

order_id  order_date customer_name channel       product  qty  unit_price  \
0     A1001  2024-01-01           Amy     web       AirPods    1      5990.0   
1     A1002  2024/01/01           Ben     app   iPhone Case    2       390.0   
2     A1004  2024-01-02         Debby     app         Mouse    1       890.0   
3     A1005  2024-01-03          Evan   store         Cable    3       199.0   
4     A1006  2024-01-03         Fiona     web       Monitor    1      7990.0   
5     A1007  2024-01-04          Gary     app  Laptop Stand    1       590.0   
6     A1008  2024-01-04         Henry     app           SSD    1      2490.0   
7     A1009  2024-01-05           Ivy     web       USB Hub    1         NaN   
8     A1010  2024-01-05          Jack     web        Webcam    1      1590.0   
9     A1011  2024-01-06           Ken     app     Desk Lamp    1       890.0   
10    A1012  2024-01-06          Lily     app         Chair    1      4990.0   
11    A1013  2024-01-07          Mike     web      Notebook    2       120.0   
12    A1014  2024-01-07          Nina     web           Pen   10        15.0   
13    A1015  2024-01-08         Oscar     app       Monitor    1      7990.0   
14    A1015  2024-01-08         Oscar     app       Monitor    1      7990.0   
15    A1016  2024-01-08           Pat     app     Gift Card    1      1000.0   
16    A1017  2024-13-01         Queen     web      Keyboard    1      1290.0   
17    A1018         NaN           Ray     app         Mouse    1       890.0   
18    A1019  2024-01-09           Sam     web      Keyboard    1      1290.0   
19    A1020  2024-01-04          Tina     app         Mouse    1       890.0   

    order_amount coupon_code        city is_member payment_status  \
0         5990.0         NaN      Taipei      TRUE           paid   
1          780.0      NEW100  New Taipei     false           paid   
2            NaN     WELCOME   Kaohsiung         0           paid   
3          597.0         NaN      Tainan      True           paid   
4         7990.0      SAVE20     Hsinchu     False           paid   
5          590.0         NaN      Taipei      TRUE           PAID   
6         2490.0         NaN      Taipei       NaN           paid   
7          690.0         NaN  New Taipei      true           paid   
8         1590.0         NaN  New Taipei      true       refunded   
9          890.0         NaN      Taipei      true           paid   
10       -4990.0         NaN      Taipei      true           paid   
11         240.0         NaN     Taoyuan      true           paid   
12         150.0         NaN      Taipei      true            NaN   
13        7990.0         NaN      Taipei      true           paid   
14        7990.0         NaN      Taipei      true           paid   
15        1000.0         NaN      Taipei      true           paid   
16        1290.0         NaN      Taipei      true           paid   
17         890.0         NaN      Taipei      true           paid   
18        1290.0         NaN      Taipei      true           paid   
19         890.0         NaN      Taipei      true           paid   

    shipping_fee                                       notes  
0           60.0                                         NaN  
1           60.0                                 first order  
2           60.0                        missing order_amount  
3           60.0                                         NaN  
4            NaN                        missing shipping_fee  
5           60.0            payment_status inconsistent case  
6           60.0                           missing is_member  
7           60.0     missing unit_price but has order_amount  
8           60.0                                         NaN  
9           60.0                                         NaN  
10          60.0  negative amount (returns?) but status paid  
11           0.0                           shipping_fee 0 ok  
12          60.0                      missing payment

In [24]:
# TODO: 顯示前 5 筆資料
df.head()

order_id  order_date customer_name channel      product  qty  unit_price  \
0    A1001  2024-01-01           Amy     web      AirPods    1      5990.0   
1    A1002  2024/01/01           Ben     app  iPhone Case    2       390.0   
2    A1004  2024-01-02         Debby     app        Mouse    1       890.0   
3    A1005  2024-01-03          Evan   store        Cable    3       199.0   
4    A1006  2024-01-03         Fiona     web      Monitor    1      7990.0   

   order_amount coupon_code        city is_member payment_status  \
0        5990.0         NaN      Taipei      TRUE           paid   
1         780.0      NEW100  New Taipei     false           paid   
2           NaN     WELCOME   Kaohsiung         0           paid   
3         597.0         NaN      Tainan      True           paid   
4        7990.0      SAVE20     Hsinchu     False           paid   

   shipping_fee                 notes  
0          60.0                   NaN  
1          60.0           first order  
2          60.0  missing order_amount  
3          60.0                   NaN  
4           NaN  missing shipping_fee

In [26]:
# TODO: 顯示後 5 筆資料
df.tail()

order_id  order_date customer_name channel    product  qty  unit_price  \
15    A1016  2024-01-08           Pat     app  Gift Card    1      1000.0   
16    A1017  2024-13-01         Queen     web   Keyboard    1      1290.0   
17    A1018         NaN           Ray     app      Mouse    1       890.0   
18    A1019  2024-01-09           Sam     web   Keyboard    1      1290.0   
19    A1020  2024-01-04          Tina     app      Mouse    1       890.0   

    order_amount coupon_code    city is_member payment_status  shipping_fee  \
15        1000.0         NaN  Taipei      true           paid          60.0   
16        1290.0         NaN  Taipei      true           paid          60.0   
17         890.0         NaN  Taipei      true           paid          60.0   
18        1290.0         NaN  Taipei      true           paid          60.0   
19         890.0         NaN  Taipei      true           paid          60.0   

                                  notes  
15                                  NaN  
16                invalid month in date  
17                   missing order_date  
18          date outside expected range  
19  late record inserted (out-of-order)

In [28]:
# TODO: 隨機顯示 5 筆資料（設定 random_state=42 確保結果一致）
df.sample(5)

order_id  order_date customer_name channel   product  qty  unit_price  \
16    A1017  2024-13-01         Queen     web  Keyboard    1      1290.0   
11    A1013  2024-01-07          Mike     web  Notebook    2       120.0   
12    A1014  2024-01-07          Nina     web       Pen   10        15.0   
8     A1010  2024-01-05          Jack     web    Webcam    1      1590.0   
10    A1012  2024-01-06          Lily     app     Chair    1      4990.0   

    order_amount coupon_code        city is_member payment_status  \
16        1290.0         NaN      Taipei      true           paid   
11         240.0         NaN     Taoyuan      true           paid   
12         150.0         NaN      Taipei      true            NaN   
8         1590.0         NaN  New Taipei      true       refunded   
10       -4990.0         NaN      Taipei      true           paid   

    shipping_fee                                       notes  
16          60.0                       invalid month in date  
11           0.0                           shipping_fee 0 ok  
12          60.0                      missing payment_status  
8           60.0                                         NaN  
10          60.0  negative amount (returns?) but status paid

In [29]:
# TODO: 顯示資料大小（幾列 x 幾欄）
# 提示：使用 df.shape
df.shape

(20, 14)

In [30]:
# TODO: 列出所有欄位名稱
# 提示：使用 df.columns
df.columns

Index(['order_id', 'order_date', 'customer_name', 'channel', 'product', 'qty',
       'unit_price', 'order_amount', 'coupon_code', 'city', 'is_member',
       'payment_status', 'shipping_fee', 'notes'],
      dtype='object')

**🤔 初步觀察：**

從上面的結果，你發現什麼問題了嗎？

（請在這裡寫下你的觀察）

---

## Part 1｜Data Profiling - 深度檢查資料

### 🎯 Exercise 1.1：檢查資料型別和基本統計

**任務：**
- 使用 `df.info()` 查看每個欄位的資料型別、非空值數量
- 使用 `df.describe(include='all')` 查看所有欄位的統計資訊

**要注意的問題：**
- 哪些欄位應該是數字但顯示為 object？
- 哪些欄位應該是日期但顯示為 object？
- 哪些欄位的非空值數量不等於總列數（代表有缺值）？

---

### 📌 提示
- 對照 learning：**Section 3「認識資料」**

In [31]:
# TODO: 使用 df.info() 檢查資料型別
df.info

<bound method DataFrame.info of    order_id  order_date customer_name channel       product  qty  unit_price  \
0     A1001  2024-01-01           Amy     web       AirPods    1      5990.0   
1     A1002  2024/01/01           Ben     app   iPhone Case    2       390.0   
2     A1004  2024-01-02         Debby     app         Mouse    1       890.0   
3     A1005  2024-01-03          Evan   store         Cable    3       199.0   
4     A1006  2024-01-03         Fiona     web       Monitor    1      7990.0   
5     A1007  2024-01-04          Gary     app  Laptop Stand    1       590.0   
6     A1008  2024-01-04         Henry     app           SSD    1      2490.0   
7     A1009  2024-01-05           Ivy     web       USB Hub    1         NaN   
8     A1010  2024-01-05          Jack     web        Webcam    1      1590.0   
9     A1011  2024-01-06           Ken     app     Desk Lamp    1       890.0   
10    A1012  2024-01-06          Lily     app         Chair    1      4990.0   
11    A1

In [32]:
# TODO: 使用 df.describe(include='all') 查看統計資訊
df.describe(include='all')

order_id  order_date customer_name channel product        qty  \
count        20          19            20      20      20  20.000000   
unique       19          11            19       3      15        NaN   
top       A1015  2024-01-04         Oscar     app   Mouse        NaN   
freq          2           3             2      11       3        NaN   
mean        NaN         NaN           NaN     NaN     NaN   1.650000   
std         NaN         NaN           NaN     NaN     NaN   2.033276   
min         NaN         NaN           NaN     NaN     NaN   1.000000   
25%         NaN         NaN           NaN     NaN     NaN   1.000000   
50%         NaN         NaN           NaN     NaN     NaN   1.000000   
75%         NaN         NaN           NaN     NaN     NaN   1.000000   
max         NaN         NaN           NaN     NaN     NaN  10.000000   

         unit_price  order_amount coupon_code    city is_member  \
count     19.000000     19.000000           3      20        19   
unique          NaN           NaN           3       6         6   
top             NaN           NaN      NEW100  Taipei      true   
freq            NaN           NaN           1      13        13   
mean    2499.157895   2018.263158         NaN     NaN       NaN   
std     2890.691118   3267.754680         NaN     NaN       NaN   
min       15.000000  -4990.000000         NaN     NaN       NaN   
25%      740.000000    643.500000         NaN     NaN       NaN   
50%     1000.000000    890.000000         NaN     NaN       NaN   
75%     3740.000000   2040.000000         NaN     NaN       NaN   
max     7990.000000   7990.000000         NaN     NaN       NaN   

       payment_status  shipping_fee        notes  
count              19     19.000000           15  
unique              3           NaN           15  
top              paid           NaN  first order  
freq               17           NaN            1  
mean              NaN     56.842105          NaN  
std               NaN     13.764944          NaN  
min               NaN      0.000000          NaN  
25%               NaN     60.000000          NaN  
50%               NaN     60.000000          NaN  
75%               NaN     60.000000          NaN  
max               NaN     60.000000          NaN

**🤔 你的發現：**

1. 哪些欄位的資料型別看起來不對？

（請在這裡寫下）

2. 有哪些欄位有缺值？

（請在這裡寫下）

---

### 🎯 Exercise 1.2：統計缺值數量

**任務：**
- 使用 `df.isna().sum()` 計算每個欄位的缺值數量
- 思考：哪些欄位的缺值最多？這對分析有什麼影響？

---

### 📌 提示
- 對照 learning：**Section 6「缺值為什麼是地雷」**

In [33]:
# TODO: 計算每個欄位的缺值數量
df.isna().sum()

order_id           0
order_date         1
customer_name      0
channel            0
product            0
qty                0
unit_price         1
order_amount       1
coupon_code       17
city               0
is_member          1
payment_status     1
shipping_fee       1
notes              5
dtype: int64

**🤔 你的分析：**

哪些欄位的缺值需要處理？為什麼？

（請在這裡寫下）

---

### 🎯 Exercise 1.3：檢查類別欄位的唯一值

**任務：**

對以下欄位使用 `.unique()` 或 `.value_counts(dropna=False)` 查看所有唯一值：
- `channel`（訂單來源）
- `is_member`（是否會員）
- `payment_status`（付款狀態）

**要注意的問題：**
- 是否有大小寫不一致？
- 是否有多種表達方式（例如：TRUE / True / true）？
- 是否有缺值（NaN）？

---

### 📌 提示
- `df['column_name'].unique()` 可以顯示所有唯一值
- `df['column_name'].value_counts(dropna=False)` 可以統計每個值的數量（包括 NaN）

In [41]:
# TODO: 檢查 channel 欄位的唯一值
df['channel'].unique()

array(['web', 'app', 'store'], dtype=object)

In [35]:
# TODO: 檢查 is_member 欄位的唯一值
df['is_member'].unique()

array(['TRUE', 'false', '0', 'True', 'False', nan, 'true'], dtype=object)

In [36]:
# TODO: 檢查 payment_status 欄位的唯一值
df['payment_status'].unique()

array(['paid', 'PAID', 'refunded', nan], dtype=object)

**🤔 你的發現：**

這些欄位有什麼格式不一致的問題？

（請在這裡寫下）

---

## Part 2｜Missing Values - 處理缺值

### 🎯 Exercise 2.1：找出特定欄位的缺值列

**任務：**

列出以下欄位有缺值的所有列：
1. `order_amount`（訂單金額）
2. `unit_price`（單價）
3. `order_date`（訂單日期）
4. `payment_status`（付款狀態）

**目的：**
- 了解缺值發生在哪些訂單
- 判斷能否從其他欄位推算出缺失的值

---

### 📌 提示
- 使用 `df[df['column_name'].isna()]` 可以篩選出該欄位為空值的列
- 對照 learning：**Section 5「loc / iloc」**

In [38]:
# TODO: 列出 order_amount 為空值的列
df[df['order_amount'].isna()]

order_id  order_date customer_name channel product  qty  unit_price  \
2    A1004  2024-01-02         Debby     app   Mouse    1       890.0   

   order_amount coupon_code       city is_member payment_status  shipping_fee  \
2           NaN     WELCOME  Kaohsiung         0           paid          60.0   

                  notes  
2  missing order_amount

In [42]:
# TODO: 列出 unit_price 為空值的列
df[df['unit_price'].isna()]

order_id  order_date customer_name channel  product  qty  unit_price  \
7    A1009  2024-01-05           Ivy     web  USB Hub    1         NaN   

   order_amount coupon_code        city is_member payment_status  \
7         690.0         NaN  New Taipei      true           paid   

   shipping_fee                                    notes  
7          60.0  missing unit_price but has order_amount

In [43]:
# TODO: 列出 order_date 為空值的列
df[df['order_date'].isna()]

order_id order_date customer_name channel product  qty  unit_price  \
17    A1018        NaN           Ray     app   Mouse    1       890.0   

    order_amount coupon_code    city is_member payment_status  shipping_fee  \
17         890.0         NaN  Taipei      true           paid          60.0   

                 notes  
17  missing order_date

In [44]:
# TODO: 列出 payment_status 為空值的列
df[df['payment_status'].isna()]

order_id  order_date customer_name channel product  qty  unit_price  \
12    A1014  2024-01-07          Nina     web     Pen   10        15.0   

    order_amount coupon_code    city is_member payment_status  shipping_fee  \
12         150.0         NaN  Taipei      true            NaN          60.0   

                     notes  
12  missing payment_status

In [45]:
df[df['coupon_code'].isna()]

order_id  order_date customer_name channel       product  qty  unit_price  \
0     A1001  2024-01-01           Amy     web       AirPods    1      5990.0   
3     A1005  2024-01-03          Evan   store         Cable    3       199.0   
5     A1007  2024-01-04          Gary     app  Laptop Stand    1       590.0   
6     A1008  2024-01-04         Henry     app           SSD    1      2490.0   
7     A1009  2024-01-05           Ivy     web       USB Hub    1         NaN   
8     A1010  2024-01-05          Jack     web        Webcam    1      1590.0   
9     A1011  2024-01-06           Ken     app     Desk Lamp    1       890.0   
10    A1012  2024-01-06          Lily     app         Chair    1      4990.0   
11    A1013  2024-01-07          Mike     web      Notebook    2       120.0   
12    A1014  2024-01-07          Nina     web           Pen   10        15.0   
13    A1015  2024-01-08         Oscar     app       Monitor    1      7990.0   
14    A1015  2024-01-08         Oscar     app       Monitor    1      7990.0   
15    A1016  2024-01-08           Pat     app     Gift Card    1      1000.0   
16    A1017  2024-13-01         Queen     web      Keyboard    1      1290.0   
17    A1018         NaN           Ray     app         Mouse    1       890.0   
18    A1019  2024-01-09           Sam     web      Keyboard    1      1290.0   
19    A1020  2024-01-04          Tina     app         Mouse    1       890.0   

    order_amount coupon_code        city is_member payment_status  \
0         5990.0         NaN      Taipei      TRUE           paid   
3          597.0         NaN      Tainan      True           paid   
5          590.0         NaN      Taipei      TRUE           PAID   
6         2490.0         NaN      Taipei       NaN           paid   
7          690.0         NaN  New Taipei      true           paid   
8         1590.0         NaN  New Taipei      true       refunded   
9          890.0         NaN      Taipei      true           paid   
10       -4990.0         NaN      Taipei      true           paid   
11         240.0         NaN     Taoyuan      true           paid   
12         150.0         NaN      Taipei      true            NaN   
13        7990.0         NaN      Taipei      true           paid   
14        7990.0         NaN      Taipei      true           paid   
15        1000.0         NaN      Taipei      true           paid   
16        1290.0         NaN      Taipei      true           paid   
17         890.0         NaN      Taipei      true           paid   
18        1290.0         NaN      Taipei      true           paid   
19         890.0         NaN      Taipei      true           paid   

    shipping_fee                                       notes  
0           60.0                                         NaN  
3           60.0                                         NaN  
5           60.0            payment_status inconsistent case  
6           60.0                           missing is_member  
7           60.0     missing unit_price but has order_amount  
8           60.0                                         NaN  
9           60.0                                         NaN  
10          60.0  negative amount (returns?) but status paid  
11           0.0                           shipping_fee 0 ok  
12          60.0                      missing payment_status  
13          60.0                       duplicate id next row  
14          60.0                               duplicate row  
15          60.0                                         NaN  
16          60.0                       invalid month in date  
17          60.0                          missing order_date  
18          60.0                 date outside expected range  
19          60.0         late record inserted (out-of-order)

---

### 🎯 Exercise 2.2：決定缺值處理策略

**任務：**

針對每個有缺值的欄位，寫下你的處理策略和理由。

**思考方向：**
- 這個欄位是否是必要的（critical）？
- 缺值能否從其他欄位計算出來？
- 填補（fillna）還是刪除（dropna）？
- 如果填補，要填什麼值？

---

### 📌 提示
- 對照 learning：**Section 7「處理缺值」**

**你的策略：**

| 欄位 | 處理方式 | 理由 |
|------|----------|------|
| order_date | （填入你的決定） | （填入你的理由） |
| order_amount | （填入你的決定） | （填入你的理由） |
| unit_price | （填入你的決定） | （填入你的理由） |
| payment_status | （填入你的決定） | （填入你的理由） |
| shipping_fee | （填入你的決定） | （填入你的理由） |
| is_member | （填入你的決定） | （填入你的理由） |

---

### 🎯 Exercise 2.3：建立清理後的 DataFrame

**任務：**

根據你的策略，逐步清理資料。

**重要：**
- 先建立 `df_clean = df.copy()`，不要直接修改原始資料
- 每個步驟寫在獨立的 cell 裡
- 每次修改後都檢查結果

**建議的處理順序：**
1. 處理 `shipping_fee`：填入預設值 60（業務規則）
2. 處理 `payment_status`：填入 'unknown'
3. 處理 `order_date`：**刪除**缺值的列（訂單日期是必要欄位）
4. 處理 `order_amount`：視情況決定刪除或重新計算

---

### 📌 提示
- 使用 `df.copy()` 複製資料
- 使用 `df['column'].fillna(value)` 填補缺值
- 使用 `df.dropna(subset=['column'])` 刪除特定欄位的缺值列

In [ ]:
# TODO: 建立 df_clean，複製原始資料
# df_clean = df.copy()

In [ ]:
# TODO: 處理 shipping_fee - 填入預設值 60
# df_clean['shipping_fee'] = df_clean['shipping_fee'].fillna(...)

In [ ]:
# TODO: 檢查 shipping_fee 還有沒有缺值

In [ ]:
# TODO: 處理 payment_status - 填入 'unknown'
# df_clean['payment_status'] = df_clean['payment_status'].fillna(...)

In [ ]:
# TODO: 檢查 payment_status 還有沒有缺值

In [ ]:
# TODO: 處理 order_date - 刪除缺值的列（訂單日期是必要欄位）
# df_clean = df_clean.dropna(subset=[...])

In [ ]:
# TODO: 檢查刪除後還剩幾筆資料
# print(f"原始資料：{len(df)} 筆")
# print(f"清理後：{len(df_clean)} 筆")

In [ ]:
# TODO: 處理 order_amount 的缺值
# 提示：觀察 A1004 這筆，order_amount 是空的，但有 qty 和 unit_price
# 可以考慮：計算 qty * unit_price，或是直接刪除
# 學生自己決定

---

## Part 3｜Type Conversion - 資料型別轉換

### 🎯 Exercise 3.1：轉換日期格式

**任務：**

將 `order_date` 從文字（object）轉換成日期格式（datetime）。

**步驟：**
1. 使用 `pd.to_datetime()` 轉換
2. 加上 `errors='coerce'` 參數，讓無效的日期變成 NaT（Not a Time）
3. 統計有多少個無效日期
4. 刪除無效日期的列

**要注意：**
- 這份資料有日期格式不一致的問題（例如：2024-01-01 vs 2024/01/01）
- 也有邏輯錯誤的日期（例如：2024-13-01，13 月不存在）

---

### 📌 提示
- `pd.to_datetime(df['column'], errors='coerce')` 會把無效日期變成 NaT
- `df['column'].isna().sum()` 可以統計 NaT 的數量

In [ ]:
# TODO: 將 order_date 轉換成 datetime 格式
# df_clean['order_date'] = pd.to_datetime(df_clean['order_date'], errors='coerce')

In [ ]:
# TODO: 統計有多少個無效日期（NaT）

In [ ]:
# TODO: 列出日期為無效的列（看看是哪些訂單）

In [ ]:
# TODO: 刪除日期無效的列
# df_clean = df_clean.dropna(subset=['order_date'])

In [ ]:
# TODO: 確認 order_date 的資料型別已經變成 datetime64
# df_clean.info()

---

### 🎯 Exercise 3.2：轉換數值格式

**任務：**

將以下欄位從文字（object）轉換成數值（numeric）：
- `qty`（數量）
- `unit_price`（單價）
- `order_amount`（訂單金額）
- `shipping_fee`（運費）

**步驟：**
1. 使用 `pd.to_numeric()` 轉換
2. 加上 `errors='coerce'` 參數，讓無法轉換的值變成 NaN
3. 檢查轉換後的結果

**要注意：**
- 這份資料有一些數字欄位被當成文字儲存
- 例如 A1003 的 order_amount 是 "1,290"（有逗號）

---

### 📌 提示
- `pd.to_numeric(df['column'], errors='coerce')` 會把無法轉換的值變成 NaN

In [ ]:
# TODO: 將 qty 轉換成數值格式
# df_clean['qty'] = pd.to_numeric(df_clean['qty'], errors='coerce')

In [ ]:
# TODO: 將 unit_price 轉換成數值格式

In [ ]:
# TODO: 將 order_amount 轉換成數值格式

In [ ]:
# TODO: 將 shipping_fee 轉換成數值格式

In [ ]:
# TODO: 檢查這些欄位的資料型別
# df_clean[['qty', 'unit_price', 'order_amount', 'shipping_fee']].dtypes

In [ ]:
# TODO: 檢查轉換後產生的新缺值
# df_clean[['qty', 'unit_price', 'order_amount', 'shipping_fee']].isna().sum()

**🤔 思考：**

為什麼 order_amount 會有新的 NaN？

提示：看看 A1003 這筆訂單的 notes

（請在這裡寫下你的想法）

---

## Part 4｜String Cleaning - 文字欄位清理

### 🎯 Exercise 4.1：統一 channel 欄位的大小寫

**任務：**

將 `channel` 欄位全部轉換成小寫。

**為什麼要做這件事？**
- 資料裡有 `web`, `WEB`, `Web` 等不同寫法
- 這些其實是同一個東西，但 pandas 會當成不同的值
- 統一成小寫後，分析才會正確

---

### 📌 提示
- 使用 `df['column'].str.lower()` 可以轉換成小寫

In [ ]:
# TODO: 先看看 channel 有哪些不同的值
# df_clean['channel'].value_counts()

In [ ]:
# TODO: 將 channel 轉換成小寫
# df_clean['channel'] = df_clean['channel'].str.lower()

In [ ]:
# TODO: 再次檢查 channel 的值（應該只剩 web, app, store）

---

### 🎯 Exercise 4.2：統一 payment_status 欄位的大小寫

**任務：**

將 `payment_status` 欄位全部轉換成小寫。

**同樣的原因：**
- 資料裡有 `paid`, `PAID`, `Paid` 等不同寫法
- 要統一格式才能正確分析

In [ ]:
# TODO: 先看看 payment_status 有哪些不同的值

In [ ]:
# TODO: 將 payment_status 轉換成小寫

In [ ]:
# TODO: 再次檢查 payment_status 的值

---

### 🎯 Exercise 4.3：去除空白字元

**任務：**

對以下欄位使用 `.str.strip()` 去除前後空白：
- `customer_name`（客戶姓名）
- `city`（城市）

**為什麼要做這件事？**
- 有些資料可能有多餘的空白（例如：" Amy " 和 "Amy" 應該是同一個人）
- strip() 可以去除前後空白

---

### 📌 提示
- `df['column'].str.strip()` 可以去除前後空白

In [ ]:
# TODO: 去除 customer_name 的前後空白
# df_clean['customer_name'] = df_clean['customer_name'].str.strip()

In [ ]:
# TODO: 去除 city 的前後空白

---

## Part 5｜Duplicates & Logic Checks - 重複與邏輯檢查

### 🎯 Exercise 5.1：找出重複的訂單

**任務：**

找出 `order_id` 重複的列。

**步驟：**
1. 使用 `df.duplicated(subset=['order_id'])` 找出重複的列
2. 列出所有重複的訂單
3. 決定要保留哪一筆（第一筆或最後一筆）
4. 刪除重複的列

---

### 📌 提示
- `df.duplicated(subset=['column'])` 可以找出重複的列
- `df.drop_duplicates(subset=['column'], keep='first')` 可以刪除重複，保留第一筆
- `keep='last'` 可以保留最後一筆

In [ ]:
# TODO: 找出 order_id 重複的列
# df_clean[df_clean.duplicated(subset=['order_id'], keep=False)]

In [ ]:
# TODO: 刪除重複的 order_id，保留第一筆
# df_clean = df_clean.drop_duplicates(subset=['order_id'], keep='first')

In [ ]:
# TODO: 確認還有沒有重複
# df_clean['order_id'].duplicated().sum()

---

### 🎯 Exercise 5.2：偵測負數金額（退款訂單）

**任務：**

找出 `order_amount` 為負數的訂單，並建立一個新欄位 `is_refund` 來標記。

**步驟：**
1. 找出 order_amount < 0 的列
2. 建立新欄位 `is_refund`（True = 退款，False = 正常訂單）

**思考：**
- 為什麼 A1012 的金額是負數，但 payment_status 是 paid？
- 這是資料錯誤，還是業務邏輯（退款）？

---

### 📌 提示
- `df['new_column'] = df['column'] < 0` 可以建立布林欄位

In [ ]:
# TODO: 找出 order_amount 為負數的列
# df_clean[df_clean['order_amount'] < 0]

In [ ]:
# TODO: 建立 is_refund 欄位
# df_clean['is_refund'] = df_clean['order_amount'] < 0

In [ ]:
# TODO: 統計有多少筆退款訂單
# df_clean['is_refund'].sum()

---

### 🎯 Exercise 5.3：偵測邏輯矛盾

**任務：**

找出 `unit_price` 是空值，但 `order_amount` 有值的列。

**思考：**
- 這種情況合理嗎？
- 能否從其他欄位推算出 unit_price？
- 該如何處理這些列？

---

### 📌 提示
- 使用多個條件：`df[(條件1) & (條件2)]`

In [ ]:
# TODO: 找出 unit_price 是空值但 order_amount 有值的列
# df_clean[(df_clean['unit_price'].isna()) & (df_clean['order_amount'].notna())]

**🤔 你的判斷：**

這些列該如何處理？

（請在這裡寫下你的想法）

---

### 🎯 Exercise 5.4：處理 A1003 的逗號分隔問題

**任務：**

A1003 這筆訂單的 `order_amount` 原本是 "1,290"（有逗號），轉換成數值時變成了 NaN。

**你有兩個選擇：**
1. 手動修正：把 order_amount 改成 1290
2. 刪除這筆訂單

**提示（不給完整解答）：**
- 如果要手動修正，可以用 `df.loc[條件, '欄位'] = 新值`
- 如果要刪除，可以用 `df.drop(index=...)`
- 學生自己決定

---

### 📌 提示
- 可以用 `df[df['order_id'] == 'A1003']` 找到這筆訂單
- 對照 learning：**Section 5「loc / iloc」**

In [ ]:
# TODO: 先找到 A1003 這筆訂單
# df_clean[df_clean['order_id'] == 'A1003']

In [ ]:
# TODO: 決定如何處理（修正或刪除）
# 學生自己決定

---

## Part 6｜Checkpoint - 最終檢查

### 🎯 Exercise 6.1：確認關鍵欄位沒有缺值

**任務：**

確認以下關鍵欄位沒有缺值（NaN）：
- `order_date`
- `order_amount`
- `payment_status`

**如果還有缺值，請回去處理。**

In [ ]:
# TODO: 檢查關鍵欄位的缺值
# df_clean[['order_date', 'order_amount', 'payment_status']].isna().sum()

---

### 🎯 Exercise 6.2：確認資料型別正確

**任務：**

確認以下欄位的資料型別：
- `order_date` 應該是 datetime64
- `qty`, `unit_price`, `order_amount`, `shipping_fee` 應該是 float 或 int
- `channel`, `payment_status` 應該是 object（文字）

In [ ]:
# TODO: 檢查資料型別
# df_clean.dtypes

---

### 🎯 Exercise 6.3：統計清理結果

**任務：**

比較清理前後的資料：
- 原始資料有幾筆？
- 清理後有幾筆？
- 刪除了多少筆？
- 刪除率是多少？

In [ ]:
# TODO: 統計清理結果
# print(f"原始資料：{len(df)} 筆")
# print(f"清理後：{len(df_clean)} 筆")
# print(f"刪除：{len(df) - len(df_clean)} 筆")
# print(f"刪除率：{(len(df) - len(df_clean)) / len(df) * 100:.1f}%")

---

### 🎯 Exercise 6.4：儲存清理後的資料

**任務：**

將清理後的資料儲存成 `data/clean_orders.csv`。

**參數說明：**
- `index=False`：不要儲存 index（列編號）
- `encoding='utf-8-sig'`：支援中文（加上 BOM）

In [ ]:
# TODO: 儲存清理後的資料
# df_clean.to_csv('data/clean_orders.csv', index=False, encoding='utf-8-sig')
# print("✅ 清理後的資料已儲存至 data/clean_orders.csv")

---

## 🎉 完成！

**恭喜你完成資料清理練習！**

### 你學到了什麼？

✅ **偵測資料問題**
- 使用 info(), describe(), isna(), unique() 檢查資料

✅ **處理缺值**
- 判斷該填補還是刪除
- 使用 fillna(), dropna()

✅ **資料型別轉換**
- 使用 pd.to_datetime(), pd.to_numeric()
- 處理轉換失敗的情況（errors='coerce'）

✅ **文字清理**
- 統一大小寫（.str.lower()）
- 去除空白（.str.strip()）

✅ **邏輯檢查**
- 找出重複資料（duplicated）
- 偵測異常值（負數、邏輯矛盾）

---

## 💭 Reflection｜學習反思

**請花 5 分鐘回答以下問題：**

### 1. 這次清理資料，你覺得最困難的是什麼？
（請在這裡寫下）


### 2. 你學到哪些「公司裡真的會遇到」的資料問題？
（請在這裡寫下）


### 3. 如果再拿到一份新的髒資料，你會先做什麼？
（請在這裡寫下你的步驟）


### 4. 資料清理大概花了你多少時間？你覺得值得嗎？
（請在這裡寫下）


---

## 📚 延伸學習

**如果你還想挑戰更多：**

1. **資料驗證**
   - 寫一個函數來檢查資料是否符合規則
   - 例如：order_amount 是否等於 qty * unit_price？

2. **自動化清理**
   - 把清理步驟寫成函數
   - 下次拿到新資料時可以直接套用

3. **資料分析**
   - 用清理後的資料做分析
   - 例如：哪個通路的訂單最多？平均金額多少？

---

**記住：資料清理佔資料分析工作的 70%**

**把資料清理好，後面的分析才會正確！** 💪